external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaRDD.scala

```diff
@@ -0,0 +1,157 @@
+  */
+class KafkaRDD[
+  K: ClassTag,
+  V: ClassTag,
+  U <: Decoder[_]: ClassTag,
+  T <: Decoder[_]: ClassTag,
+  R: ClassTag](
+    sc: SparkContext,
+    val kafkaParams: Map[String, String],
+    val fromOffsets: Map[TopicAndPartition, Long],
+    val untilOffsets: Map[TopicAndPartition, Long],
+    messageHandler: MessageAndMetadata[K, V] => R
+  ) extends RDD[R](sc, Nil) with Logging {
+
+  assert(fromOffsets.keys == untilOffsets.keys,
+    "Must provide both from and until offsets for each topic/partition")
+
+  override def getPartitions: Array[Partition] = fromOffsets.zipWithIndex.map { kvi =>
+    val ((tp, from), index) = kvi
+    new KafkaRDDPartition(index, tp.topic, tp.partition, from, untilOffsets(tp))
+  }.toArray
+
+  override def compute(thePart: Partition, context: TaskContext) = {
+    val part = thePart.asInstanceOf[KafkaRDDPartition]
+    if (part.fromOffset >= part.untilOffset) {
+      log.warn("Beginning offset is same or after ending offset " +
+        s"skipping ${part.topic} ${part.partition}")
+      Iterator.empty
+    } else {
+      new NextIterator[R] {
+        context.addTaskCompletionListener{ context => closeIfNeeded() }
+
+        val kc = new KafkaCluster(kafkaParams)
+        log.info(s"Computing topic ${part.topic}, partition ${part.partition} " +
+          s"offsets ${part.fromOffset} -> ${part.untilOffset}")
+        val keyDecoder = classTag[U].runtimeClass.getConstructor(classOf[VerifiableProperties])
+          .newInstance(kc.config.props)
+          .asInstanceOf[Decoder[K]]
+        val valueDecoder = classTag[T].runtimeClass.getConstructor(classOf[VerifiableProperties])
+          .newInstance(kc.config.props)
+          .asInstanceOf[Decoder[V]]
+        val consumer: SimpleConsumer = kc.connectLeader(part.topic, part.partition).fold(
+          errs => throw new Exception(
+            s"Couldn't connect to leader for topic ${part.topic} ${part.partition}: " +
+              errs.mkString("\n")),
+          consumer => consumer
+        )
+        var requestOffset = part.fromOffset
+        var iter: Iterator[MessageAndOffset] = null
+
+        override def close() = consumer.close()
+
+        override def getNext: R = {
+          if (iter == null || !iter.hasNext) {
+            val req = new FetchRequestBuilder().
+              addFetch(part.topic, part.partition, requestOffset, kc.config.fetchMessageMaxBytes).
+              build()
+            val resp = consumer.fetch(req)
+            if (resp.hasError) {
+              val err = resp.errorCode(part.topic, part.partition)
+              if (err == ErrorMapping.LeaderNotAvailableCode ||
+                err == ErrorMapping.NotLeaderForPartitionCode) {
+                log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
+                  s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
+                Thread.sleep(kc.config.refreshLeaderBackoffMs)
```

Comment:
That's only happening in an error case, and is being done to make sure that you don't burn through the maximum number of failed stages while leader election is happening.

I'm pretty sure the existing spark kafka dstreams are doing the exact same thing, because that's what the kafka project's consumer code does in ConsumerFetcherManager.

Do you have an alternative suggestion?